### ISSUE TITLE
v1.0.0?

### ISSUE THREAD

**Issue by eventualbuddha** [rollup/rollup#769]
Rollup has a reasonably stable API, so I think we should consider cutting a 1.0.0 release and doing semver after that.

To that end, we could use something like [semantic-release](https://github.com/semantic-release/semantic-release) plus describing the type of changes in commits/PRs. I initially wasn't a fan of that approach but I've been using it for a bit in decaffeinate projects and found I quite like it. I'm sure you have your own ideas about how you'd like to handle releases, @Rich-Harris. Just my 2¢.

**Comment by Rich-Harris** [comment]
I was wondering about that. I don't have particularly strong feelings – I was thinking it'd be nice to implement some of the value tracking stuff we've occasionally talked about (which would be quite a big refactor) and fix some of the outstanding bugs (e.g. https://github.com/rollup/rollup/issues/716 and similar) before declaring we've reached version 1, and working on Bublé and Prosecco has given me a few ideas about slightly different ways to do things – but then I've no idea when I'll be able to commit time to that, and perhaps if the API isn't likely to change then there's no real harm in it.

**Comment by Rich-Harris** [comment]
As for semantic-release, on a few occasions I've thought about adding it to my projects, but I've always backed out. It seems like there's a lot to learn/do in order to automate what in reality isn't a particularly onerous task, and it seems like you need to cede a bit of control over the whole process, which is fine until you find yourself in a bit of a jam and need to duct-tape your way out of it. Maybe I'm just being old-fashioned?

**Comment by eventualbuddha** [comment]
> I was thinking it'd be nice to implement some of the value tracking stuff we've occasionally talked about (which would be quite a big refactor) and fix some of the outstanding bugs (e.g. #716 and similar) before declaring we've reached version 1

I know what you mean. There does seem to be momentum toward a world of semantic versions for libraries and [infinite versions](https://blog.codinghorror.com/the-infinite-version/) for applications. I like the idea of assigning a strict, non-emotional meaning to version numbers for libraries and tools like rollup. The npm community, in particular, seems to have embraced this approach. However, there will likely be holdouts for years to come who see a 3.0.0 release and think there must be amazing new features when really it's just that a breaking change was introduced. I certainly feel it sometimes and the statement "declaring we've reached version 1" is somewhat in that camp.

> Maybe I'm just being old-fashioned?

Maybe. I initially had the same reaction to Greenkeeper. I didn't like the idea of some tool having commit access to my repo. But then I saw how useful it was on other projects, such as rollup, and how I would no longer have to remember to periodically run `ncu -ua && npm update`, sort out what broke it, and whether I screwed something up and it worked on my machine but not in CI. It helped keep projects from getting stale (i.e. "turning brown", if you'll pardon the pun).

So I enabled it for the decaffeinate projects and I've been pretty happy so far. When I encountered semantic-release I had some of the initial reactions you mentioned. So far I haven't had to work around the system, but it has only been a few weeks. I'll let you know how I like it in a month or so. In the meantime I don't have to worry about whether I remember to cut a release after merging something or to update the website -- it just happens automatically. That's nice, since they're really not things I want to have to remember. Now I wish I had it for projects at work.

**Comment by Aqours** [comment]
+1

**Comment by eventualbuddha** [comment]
After more than a month with semantic-release, I've had two occasions where I've had to push an empty commit to get a release out. Both would have been avoidable with more forethought, but even doing the workaround wasn't a big deal.

Doing automatic releases of the website for decaffeinate has been a little trickier, but I believe I've settled on a decent system. Like with rollupjs.org, I have a separate repo for decaffeinate's website that has its own Travis build that publishes using surge.sh. It has a package.json that depends on decaffeinate, browserify, and whatever else needed to build the browser version. When decaffeinate publishes a new version, it also bumps the version in the package.json for the website which then triggers a rebuild in Travis for it too. So now it publishes a new version of the website whenever I change the website itself or when decaffeinate gets a new version.

So consider me still in favor of both a v1.0.0 for rollup and, separately, using semantic-release for managing releases.

**Comment by jhabdas** [comment]
:+1: for SemVer, because predictability. But it should be driven by a roadmap with some core needs, and perhaps beforehand the CLI and plug-in parts of the codebase pulled into separate, independently versiones packages (if it makes sense).

**Comment by jhabdas** [comment]
http://semver.org/#how-do-i-know-when-to-release-100

> If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backwards compatibility, you should probably already be 1.0.0.

It's time. Let's cut a release.

**Comment by rkirsling** [comment]
It is definitely time.

There are projects at my organization making use of `grunt-rollup`, whose latest version suddenly broke yesterday after being stable for 8 months. Admittedly, part of the blame lies with that package—by depending on `0.x`, breakage could have occurred at any time—and part of the blame lies with the consumer repo which didn't itself specify a `rollup` version. Still, you can see how this happened, when any given 0.x.0 release "might be major, might be minor".

Rollup is a fantastically popular package and has been for more than a year now.
To continue to release under a zero major is a disservice to your community.

**Comment by Rich-Harris** [comment]
@rkirsling I definitely see where you're coming from, and I really would like to get to 1.0.0 soon. The issue as I see it currently is that there are some things about Rollup that are just unfinished (see #1479 for example, though there are certainly others) and it makes sense to square those away prior to a v1 release otherwise we'll be on version 6 before we know it. For example, merging #1475 the other day would have forced a bump from 1 to 2, even though it probably affects about a dozen people. (I know the whole 'integers are free' counterpoint, but it doesn't really chime with how most people think of software, especially when the changes in question are basically just chores.)

Basically, I don't believe in marking something as 'done' when there are still lots of issues open — some of them bugs, some of them feature requests that are likely to result in breaking changes. We'll get there soon, I promise!

**Comment by BlueRaja** [comment]
> I don't believe in marking something as 'done' when there are still lots of issues open

"Bumping to 1.0" doesn't mean "being done/issue-free", or no project would ever go to 1.0.  It  should mean it's ready for production-use, which rollup already is _(as evidenced by its widespread use in production environments everywhere)_.

> we'll be on version 6 before we know it [..] it doesn't really chime with how most people think of software

But it does chime with how your users (developers) understand semantic versioning.  When you make a breaking-change, you have to bump the breaking-change number regardless; staying at 0.x only means that the _(many, unfortunately)_ developers who don't understand how 0.x versions work will have issues with rollup, like with `grunt-rollup`

**Comment by styfle** [comment]
@eventualbuddha It seems like all the linked issues/PRs are closed/merged.
I am in favor of releasing 1.0 and actually following semver.
This is very normal and practically expected for packages published to npm.
Rollup is stable and there are at least [600](https://www.npmjs.com/package/rollup) other packages that depend on it.

Let's make `1.0.0` happen! 🎉

**Comment by trusktr** [comment]
Look at Chrome, for example, it is v69, and we're all fine!

**Comment by shellscape** [comment]
@trusktr love the enthusiasm! but _please_ use the edit feature on your first reply rather than making three consecutive replies within a few minutes of each. For those of us attentive to the repo and subscribed, it's awfully spammy.

**Comment by LongTengDao** [comment]
@eventualbuddha 
Is the version 0.x semantic?
Like 0.incompatible.compatible?
I mean, can I use 0.66.x to achieve the goal like major.x?
Or the version before 1.0 could be backward-incompatible, even just change the third patch number?

**Comment by shellscape** [comment]
@LongTengDao with versions _prior_ to `1.0.0`, any minor version (`0.x.0`) may contain breaking changes, and the release notes should be consulted before updating. See: https://semver.org/#spec-item-4

**Comment by LongTengDao** [comment]
@shellscape will patch version (`0.66.x`) contain breaking changes?

**Comment by eventualbuddha** [comment]
@LongTengDao yes, patch versions could contain breaking changes in a `v0.x.x` version:

> Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.

**Comment by alexanderby** [comment]
![homer-simpson-semver](https://user-images.githubusercontent.com/9638588/49880722-cec0c480-fe3d-11e8-959a-009bb64dcd78.jpg)

**Comment by jedwards1211** [comment]
**Using 0.x.x is a cardinal sin**.
Developers suffer when it's harder for them to tell if breaking changes have occurred.

**Comment by jedwards1211** [comment]
> For example, merging #1475 the other day would have forced a bump from 1 to 2, even though it probably affects about a dozen people.

Well, the bump from 1 to 2 **is exactly what you should do**, because it's *the kindest thing you can do to users*.  That way, people will know that breaking changes have occurred, will be able to quickly find what the breaking changes are in the releases/changelog, and will be relieved that they're not one of the half dozen people affected.
I'm trying to upgrade an OSS project that's stuck on `rollup` 0.44 and you have forced me to hunt for breaking changes in over 23 versions in the changelog (0.44-0.67).  If you were using major versioning there would be a lot fewer places for me to look for breaking changes.

**Comment by lukastaegert** [comment]
Guys, calm down. Even though I am doing it basically on my own right now, we are finally really close to the 1.0 now: https://github.com/rollup/rollup/projects/1

At this point it is basically a few reviews, one last 0.x release with the PRs currently waiting to be merged and then a 1.0, maybe even before Christmas if I manage to update the docs until then. Help, of course, is always welcome.

**Comment by shellscape** [comment]
> I'm trying to upgrade an OSS project that's stuck on rollup 0.44 and you have forced me to hunt for breaking changes in over 23 versions in the changelog (0.44-0.67).

There is some amount of responsibility on the user's end when using a module that has not hit semver 1.0.0. For whatever reason, either the project you inherited or the project you've been maintaining has not chosen to keep up to date, and the onus for that falls on the shoulders of the person(s) who had _made the choice_ to use a module in a 0.x track, and who has consistently made the decision not to upgrade _for 23 straight version releases_.

Directing your frustrations at a free, open source project and its volunteer maintainers is not right. Rather, it would seem your frustrations should be focused _inward_. Criticisms such as the above serve to exemplify the gripes that FOSS maintainers have with an ever-increasing ungrateful user base.

@lukastaegert is working his butt off in his spare time to get the project to a point where all maintainers agree that it's ready for 1.0.0. He's carrying the load right now because the rest of us have other IRL things we must take care of. While you may have carved out the time for 7,000 Github contributions, we are beholden to our career and offline responsibilities.

Patience, gratitude, and understanding are the appropriate expressions for replying to this thread.

**Comment by eventualbuddha** [comment]
Congrats on 1.0! 🎉
